File: hoffman-agents-python/src/conscious_agent/agent/decision_map.py
from __future__ import annotations

import random
from typing import Literal

import numpy as np

from .experience_space import ExperienceSpace

OutputState = Literal["core", "lexicon", "explore", "idle"]

CORE_TOKENS = ["I", "notice", "familiar", "different", "wait"]

SIGNATURE_MATCH_THRESHOLD = 0.3


def _build_transition_signature(
    prev_id: int | None,
    curr_id: int,
    embedding_dim: int,
) -> np.ndarray:
    sig = np.zeros(embedding_dim, dtype=np.float64)
    if prev_id is not None:
        combined = hash(f"{prev_id}->{curr_id}")
        sig[combined % embedding_dim] = 1.0
    sig[curr_id % embedding_dim] = 1.0
    norm = np.linalg.norm(sig)
    if norm > 0:
        sig = sig / norm
    return sig


def _sample_lexicon_label(experience: ExperienceSpace, vocab_size: int = 5,
                          rng: random.Random = random._inst) -> str | None:
    entries = experience.lexicon.sorted_by_integration()
    if not entries:
        return None

    meta_id = experience.meta_trie.last_meta_state
    if meta_id is not None:
        predicted = experience.meta_trie.predict_token(meta_id, min_observations=3)
        if predicted is not None:
            for e in entries:
                if e.output_token == predicted:
                    e.encounter_count += 1
                    e.integration_depth = min(e.integration_depth + 0.02, 1.0)
                    return predicted

    curr_id = experience.last_world_state_id
    if curr_id is not None:
        prev_id = None
        recent = experience.trace_buffer.get_recent(2)
        if len(recent) >= 2:
            prev_id = recent[-2].to_state
        query_sig = _build_transition_signature(prev_id, curr_id, experience.lexicon.embedding_dim)
        best_label, best_dist = experience.lexicon.nearest_label(query_sig)
        if best_label and best_dist < SIGNATURE_MATCH_THRESHOLD:
            entry = experience.lexicon.lookup_by_label(best_label)
            if entry is not None and entry.integration_depth > 0.01:
                entry.encounter_count += 1
                entry.integration_depth = min(entry.integration_depth + 0.02, 1.0)
                return entry.output_token

    weighted = []
    for e in entries:
        w = max(e.integration_depth, 0.01)
        w *= (1.0 + 0.1 * e.encounter_count)
        if e.labeling_source == "adopted":
            w *= 3.0
        weighted.append((w, e))
    weighted.sort(key=lambda x: -x[0])
    top_n = weighted[:vocab_size]
    weights = [w for w, _ in top_n]
    total = sum(weights)
    r = rng.random() * total
    cumulative = 0.0
    for w, entry in top_n:
        cumulative += w
        if r <= cumulative:
            return entry.output_token
    return top_n[-1][1].output_token if top_n else None


def _next_state(current: OutputState, p_stable: float, p_lexicon: float,
                p_explore: float,
                rng: random.Random = random._inst) -> OutputState:
    r = rng.random()
    if current == "lexicon":
        if r < 0.70:
            return "core"
        r -= 0.70
        if r < 0.15:
            return "lexicon"
        r -= 0.15
        if r < 0.10:
            return "explore"
        return "idle"
    else:
        if r < p_stable:
            return "core"
        r -= p_stable
        if r < p_lexicon:
            return "lexicon"
        r -= p_lexicon
        if r < p_explore:
            return "explore"
        return "idle"


def decide(
    experience: ExperienceSpace,
    max_tokens: int = 8,
    p_stable: float = 0.80,
    p_lexicon: float = 0.10,
    p_explore: float = 0.05,
    ergodic_state: OutputState | None = None,
) -> tuple[list[str], OutputState]:
    if not experience.self_token.is_locked():
        return (["wait"], "idle")

    state = ergodic_state if ergodic_state is not None else "core"
    next_state_val = _next_state(state, p_stable, p_lexicon, p_explore)

    if next_state_val == "core":
        tokens = [experience.self_token.token, "notice"]
        snapshot = experience.meta_trie.get_meta_state_snapshot(
            experience.meta_trie.last_meta_state
        ) if experience.meta_trie.last_meta_state is not None else None
        tokens.append(experience.self_token.token)
        if snapshot is not None and snapshot.mean_prediction_error > 0.3:
            tokens.append("different")
        else:
            tokens.append("familiar")

    elif next_state_val == "lexicon":
        label = _sample_lexicon_label(experience)
        if label is not None:
            tokens = [experience.self_token.token, "notice", label]
        else:
            tokens = [experience.self_token.token, "notice", "familiar"]

    elif next_state_val == "explore":
        token = random.choice(CORE_TOKENS)
        tokens = [token]

    else:
        tokens = ["wait"]

    return (tokens[:max_tokens], next_state_val)

File: hoffman-agents-react-native/src/agent/decision-map.js
const CORE_TOKENS = ['I', 'notice', 'familiar', 'different', 'wait'];
const SIGNATURE_MATCH_THRESHOLD = 0.3;

function _buildTransitionSignature(prevId, currId, embeddingDim) {
  const sig = new Float64Array(embeddingDim);
  if (prevId !== null && prevId !== undefined) {
    const h = Math.abs(hashCode(`${prevId}->${currId}`));
    sig[h % embeddingDim] = 1;
  }
  sig[currId % embeddingDim] = 1;
  let norm = 0;
  for (let i = 0; i < sig.length; i++) norm += sig[i] * sig[i];
  norm = Math.sqrt(norm);
  if (norm > 0) for (let i = 0; i < sig.length; i++) sig[i] /= norm;
  return sig;
}

function hashCode(str) {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    hash = ((hash << 5) - hash) + str.charCodeAt(i);
    hash |= 0;
  }
  return hash;
}

const _random = () => Math.random();

function _sampleLexiconLabel(experience, vocabSize = 5, rng = _random) {
  const entries = experience.lexicon.sortedByIntegration();
  if (entries.length === 0) return null;

  const metaId = experience.metaTrie.lastMetaState;
  if (metaId !== null) {
    const predicted = experience.metaTrie.predictToken(metaId, 3);
    if (predicted) {
      for (const e of entries) {
        if (e.outputToken === predicted) {
          e.encounterCount++;
          e.integrationDepth = Math.min(e.integrationDepth + 0.02, 1);
          return predicted;
        }
      }
    }
  }

  const currId = experience.lastWorldStateId;
  if (currId !== null) {
    let prevId = null;
    const recent = experience.traceBuffer.getRecent(2);
    if (recent.length >= 2) prevId = recent[recent.length - 2].toState;
    const querySig = _buildTransitionSignature(prevId, currId, experience.lexicon._embeddingDim);
    const [bestLabel, bestDist] = experience.lexicon.nearestLabel(querySig);
    if (bestLabel && bestDist < SIGNATURE_MATCH_THRESHOLD) {
      const entry = experience.lexicon.lookupByLabel(bestLabel);
      if (entry && entry.integrationDepth > 0.01) {
        entry.encounterCount++;
        entry.integrationDepth = Math.min(entry.integrationDepth + 0.02, 1);
        return entry.outputToken;
      }
    }
  }

  const weighted = entries.map(e => {
    let w = Math.max(e.integrationDepth, 0.01);
    w *= (1 + 0.1 * e.encounterCount);
    if (e.labelingSource === 'adopted') w *= 3;
    return [w, e];
  });
  weighted.sort((a, b) => b[0] - a[0]);
  const topN = weighted.slice(0, vocabSize);
  const total = topN.reduce((s, [w]) => s + w, 0);
  let r = rng() * total, cumulative = 0;
  for (const [w, entry] of topN) {
    cumulative += w;
    if (r <= cumulative) return entry.outputToken;
  }
  return topN.length > 0 ? topN[topN.length - 1][1].outputToken : null;
}

function _nextState(current, pStable, pLexicon, pExplore, rng = _random) {
  const r = rng();
  if (current === 'lexicon') {
    if (r < 0.7) return 'core';
    if (r < 0.85) return 'lexicon';
    if (r < 0.95) return 'explore';
    return 'idle';
  } else {
    if (r < pStable) return 'core';
    if (r < pStable + pLexicon) return 'lexicon';
    if (r < pStable + pLexicon + pExplore) return 'explore';
    return 'idle';
  }
}

function decide(experience, {
  maxTokens = 8, pStable = 0.8, pLexicon = 0.1, pExplore = 0.05, ergodicState = null, rng = _random
} = {}) {
  if (!experience.selfToken.isLocked()) {
    return [['wait'], 'idle'];
  }

  const state = ergodicState || 'core';
  const next = _nextState(state, pStable, pLexicon, pExplore, rng);

  let tokens;
  if (next === 'core') {
    tokens = [experience.selfToken.token, 'notice'];
    const snapshot = experience.metaTrie.lastMetaState !== null
      ? experience.metaTrie.getMetaStateSnapshot(experience.metaTrie.lastMetaState) : null;
    tokens.push(experience.selfToken.token);
    tokens.push(snapshot && snapshot.meanPredictionError > 0.3 ? 'different' : 'familiar');
  } else if (next === 'lexicon') {
    const label = _sampleLexiconLabel(experience, 5, rng);
    tokens = [experience.selfToken.token, 'notice', label || 'familiar'];
  } else if (next === 'explore') {
    tokens = [CORE_TOKENS[Math.floor(rng() * CORE_TOKENS.length)]];
  } else {
    tokens = ['wait'];
  }

  return [tokens.slice(0, maxTokens), next];
}

module.exports = { decide, _sampleLexiconLabel, CORE_TOKENS };
